### DISCUSSION TITLE
Mitigating the unique ID in GraphQL spec with a table-level ID uniqueness

### DISCUSSION THREAD

**Discussion by amirmi** [graphql-java/graphql-java#2725] [Q&A]
We've built a GraphQL server communicating with a legacy REST-based backend, on top of a MySQL DB where the IDs for each row are unique for tables. so we have cases where (e.g.) a User and Umbrella, have the same ID. 

Then we load them both into the client with Relay, and of course, boom. 

Is there a way to define something on the GraphQL side to have "ID", become a composite of [typename]:[id] on the way out, and the reverse when coming back through queries and mutations? 

(the 'reverse', i.e. composite to type + ID, could also help us implement the 'node(id)' root query)

We're are defining the schema using SDL and also using https://github.com/graphql-java-generator/graphql-java-generator, for generating the skeletons for the field implementations.

**Comment by bbakerman** [comment]
I cant answer on what this graphql-java-generator does

But you need a layer in your datafetchers that indeed encodes `[typeofobject : id]` into ID fields on the way out and interprets them in the way in.

In plain old graphql-java you would do this in the DataFethcers .  But I dont know this generator framwork

what makes up and `ID` is 100% in your court.  graphql-java is agnostic to the data

**Comment by jord1e** [comment]
> Is there a way to define something on the GraphQL side to have "ID", become a composite of [typename]:[id] on the way out, and the reverse when coming back through queries and mutations?

Not for mutations, you can create a utility class to work with Relay-like global ids.

You may override the `id` **field datafetcher** for certain types:

![afbeelding](https://user-images.githubusercontent.com/30464310/154715190-65c59df9-6db7-4c6d-acc9-4abb80fc6469.png)

This makes the T1.id field a composite in the form of `base64url("T1:" + t1.a())` when being queried
